### PR TITLE
Produce debuggable output when a kernel 'make' command fails

### PR DESCRIPTION
**Pull request by stefwalter**
Previously the returncode of kernel builds and the output was
not considered and/or hidden. This made it hard to debug what
was going on and why a build failed.

### CODE REVIEW

**Comment by stefwalter** [comment]
Previous output:

```
# skt -vvv build
2018-03-26 13:57:15,339     INFO   basecfg: /var/config/kernel-x86_64.config
2018-03-26 13:57:15,339     INFO   cfgtype: olddefconfig
2018-03-26 13:57:15,339     INFO   prepare config: ['make', '-C', '/var/work', 'olddefconfig']
make: Entering directory '/var/work'
make[1]: Entering directory '/var/build'
  HOSTCC  scripts/basic/fixdep
  GEN     ./Makefile
  HOSTCC  scripts/kconfig/conf.o
  YACC    scripts/kconfig/zconf.tab.c
  LEX     scripts/kconfig/zconf.lex.c
  HOSTCC  scripts/kconfig/zconf.tab.o
  HOSTLD  scripts/kconfig/conf
scripts/kconfig/conf  --olddefconfig Kconfig
#
# configuration written to .config
#
make[1]: Leaving directory '/var/build'
make: Leaving directory '/var/work'
2018-03-26 13:57:17,336     INFO   building kernel: ['make', '-C', '/var/work', 'INSTALL_MOD_STRIP=1', '-j8', 'targz-pkg']
Traceback (most recent call last):
  File "/usr/bin/skt", line 11, in <module>
    load_entry_point('skt==0.0.1', 'console_scripts', 'skt')()
  File "/usr/lib/python2.7/site-packages/skt/executable.py", line 607, in main
    args.func(cfg)
  File "/usr/lib/python2.7/site-packages/skt/executable.py", line 105, in wrapper
    func(cfg)
  File "/usr/lib/python2.7/site-packages/skt/executable.py", line 179, in cmd_build
    raise e
Exception: Failed to find tgz path in stdout
```

After the change:

```
# skt -vvv build
2018-03-26 13:45:03,252     INFO   basecfg: /var/config/kernel-x86_64.config
2018-03-26 13:45:03,253     INFO   cfgtype: olddefconfig
2018-03-26 13:45:03,271     INFO   prepare config: ['make', '-C', '/var/work', 'olddefconfig']
make: Entering directory '/var/work'
make[1]: Entering directory '/var/build'
  HOSTCC  scripts/basic/fixdep
  GEN     ./Makefile
  HOSTCC  scripts/kconfig/conf.o
  YACC    scripts/kconfig/zconf.tab.c
  LEX     scripts/kconfig/zconf.lex.c
  HOSTCC  scripts/kconfig/zconf.tab.o
  HOSTLD  scripts/kconfig/conf
scripts/kconfig/conf  --olddefconfig Kconfig
#
# configuration written to .config
#
make[1]: Leaving directory '/var/build'
make: Leaving directory '/var/work'
2018-03-26 13:45:30,061     INFO   building kernel: ['make', '-C', '/var/work', 'INSTALL_MOD_STRIP=1', '-j32', 'targz-pkg']
make: Entering directory '/var/work'
make[1]: Entering directory '/var/build'
  GEN     ./Makefile
scripts/kconfig/conf  --silentoldconfig Kconfig
  CHK     include/config/kernel.release
  UPD     include/config/kernel.release
make KBUILD_SRC=
  CHK     include/config/kernel.release
  HOSTCC  scripts/basic/bin2c
  GEN     ./Makefile
  CHK     include/generated/uapi/linux/version.h
  UPD     include/generated/uapi/linux/version.h
  SYSTBL  arch/x86/include/generated/asm/syscalls_32.h
  SYSHDR  arch/x86/include/generated/asm/unistd_32_ia32.h
  SYSTBL  arch/x86/include/generated/asm/syscalls_64.h
  SYSHDR  arch/x86/include/generated/asm/unistd_64_x32.h
  HYPERCALLS arch/x86/include/generated/asm/xen-hypercalls.h
  SYSHDR  arch/x86/include/generated/uapi/asm/unistd_32.h
  SYSHDR  arch/x86/include/generated/uapi/asm/unistd_64.h
  SYSHDR  arch/x86/include/generated/uapi/asm/unistd_x32.h
  CHK     include/generated/utsrelease.h
  UPD     include/generated/utsrelease.h
  WRAP    arch/x86/include/generated/uapi/asm/bpf_perf_event.h
  WRAP    arch/x86/include/generated/uapi/asm/poll.h
  DESCEND  objtool
  Using /var/work as source for kernel
  /var/work is not clean, please run 'make mrproper'
  in the '/var/work' directory.
  HOSTCC  arch/x86/tools/relocs_32.o
  HOSTCC  arch/x86/tools/relocs_64.o
  HOSTCC  arch/x86/tools/relocs_common.o
make[5]: *** [/var/work/Makefile:1086: prepare3] Error 1
make[5]: *** Waiting for unfinished jobs....
  WRAP    arch/x86/include/generated/asm/dma-contiguous.h
  WRAP    arch/x86/include/generated/asm/early_ioremap.h
  WRAP    arch/x86/include/generated/asm/mcs_spinlock.h
  WRAP    arch/x86/include/generated/asm/mm-arch-hooks.h
  HOSTCC   /var/build/tools/objtool/fixdep.o
  HOSTLD   /var/build/tools/objtool/fixdep-in.o
  LINK     /var/build/tools/objtool/fixdep
  HOSTLD  arch/x86/tools/relocs
  MKDIR    /var/build/tools/objtool/arch/x86/lib/
  GEN      /var/build/tools/objtool/arch/x86/lib/inat-tables.c
  CC       /var/build/tools/objtool/builtin-check.o
  CC       /var/build/tools/objtool/exec-cmd.o
  CC       /var/build/tools/objtool/help.o
  CC       /var/build/tools/objtool/pager.o
  CC       /var/build/tools/objtool/builtin-orc.o
  CC       /var/build/tools/objtool/parse-options.o
  CC       /var/build/tools/objtool/arch/x86/decode.o
  CC       /var/build/tools/objtool/run-command.o
  CC       /var/build/tools/objtool/sigchain.o
  CC       /var/build/tools/objtool/check.o
  CC       /var/build/tools/objtool/orc_gen.o
  CC       /var/build/tools/objtool/subcmd-config.o
  CC       /var/build/tools/objtool/orc_dump.o
  CC       /var/build/tools/objtool/elf.o
  CC       /var/build/tools/objtool/special.o
  CC       /var/build/tools/objtool/objtool.o
  CC       /var/build/tools/objtool/libstring.o
  CC       /var/build/tools/objtool/str_error_r.o
  LD       /var/build/tools/objtool/arch/x86/objtool-in.o
  LD       /var/build/tools/objtool/libsubcmd-in.o
  AR       /var/build/tools/objtool/libsubcmd.a
  LD       /var/build/tools/objtool/objtool-in.o
  LINK     /var/build/tools/objtool/objtool
make[4]: *** [Makefile:146: sub-make] Error 2
make[3]: *** [Makefile:24: __sub-make] Error 2
make[2]: *** [/var/work/scripts/package/Makefile:120: targz-pkg] Error 2
make[1]: *** [/var/work/Makefile:1398: targz-pkg] Error 2
make[1]: Leaving directory '/var/build'
make: *** [Makefile:146: sub-make] Error 2
make: Leaving directory '/var/work'
Traceback (most recent call last):
  File "/usr/bin/skt", line 11, in <module>
    load_entry_point('skt==0.0.1', 'console_scripts', 'skt')()
  File "/usr/lib/python2.7/site-packages/skt/executable.py", line 607, in main
    args.func(cfg)
  File "/usr/lib/python2.7/site-packages/skt/executable.py", line 105, in wrapper
    func(cfg)
  File "/usr/lib/python2.7/site-packages/skt/executable.py", line 179, in cmd_build
    raise e
Exception: Kernel build failed: make -C /var/work INSTALL_MOD_STRIP=1 -j32 targz-pkg

```

File: skt/__init__.py [hunk 1]
```diff
@@ -482,6 +483,10 @@ def getrelease(self):
         args = self.defmakeargs + ["kernelrelease"]
         mk = subprocess.Popen(args, stdout=subprocess.PIPE)
         (stdout, stderr) = mk.communicate()
+        if mk.returncode != 0:
+            raise Exception("Kernel release lookup failed: ".format(
```
[the remark this answers is not in the document]
Comment (stefwalter) [reply skt/__init__.py]: Thanks fixed!

**Comment by stefwalter** [comment]
Added "command" word and use of ```select``` to see progress in real time.

File: skt/__init__.py [hunk 2]
```diff
@@ -506,7 +512,24 @@ def mktgz(self, clean=True):
...
+        while stdout_fd > -1:
+            ret = select.select([stdout_fd], [], [], 10)
+            for fd in ret[0]:
+                if fd == stdout_fd:
+                    data = os.read(fd, 1024)
+                    if not data:
+                        stdout_fd = -1
+                        mk.stdout.close()
+                    sys.stdout.write(data)
+                    stdout += data.decode('utf-8', 'replace')
```
Comment (spbnick) [inline skt/__init__.py]: Hmm, why do we need to use `select` for reading from a single file descriptor and not doing anything else? I would have thought you can just read it in a loop, can't you?

[the remark this answers is not in the document]
Comment (stefwalter) [reply skt/__init__.py]: So this code in python3 ready.

File: skt/__init__.py [hunk 3]
```diff
@@ -506,7 +512,24 @@ def mktgz(self, clean=True):
         mk = subprocess.Popen(args,
                               stdout=subprocess.PIPE,
                               stderr=subprocess.STDOUT)
-        (stdout, stderr) = mk.communicate()
```
[the remark this answers is not in the document]
Comment (stefwalter) [reply skt/__init__.py]: @agustinhenze Could you please turn your commit into a pull request? I can then close this one. 

To be honest I don't care about the minutia of the technique used ... I just want to not see the build.log appear in one shot.

**Comment by stefwalter** [comment]
@agustinhenze seems to want to take this pull request over. Thanks! Could you make a new one with the code you pasted here? The approach seems fine to me.

**Comment by spbnick** [comment]
@stefwalter, only saw your comment about `select` just now. Ah, you mean that you don't want to wait for any buffer to fill up and spit out the output ASAP? Yeah, good idea, thanks.

File: skt/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (agustinhenze) [reply skt/__init__.py]: I didn't want to bother with the purity of the technique, sorry. I have suffered in the past trying to control processes from python and it's really painful. Let me write some test cases and then I'll open the PR.

**Comment by stefwalter** [comment]
Replaced with #40

**Pull request closed without merging**